File: lib/rank.js
function norm(s){return (s||'').toLowerCase()}
function has(text,term){return norm(text).includes(norm(term))}
export function rankStory(story,profile){const text=`${story.title} ${story.summary||''}`;let topic=0,region=0,entity=0,quality=0,noise=0;const topics=[];
 for(const [k,w] of Object.entries(profile.topics)){const aliases=profile.topicAliases[k]||[k];if(aliases.some(a=>has(text,a))){topics.push(k);topic+=w*10}}
 topic=Math.min(30,topic);
 for(const [r,w] of Object.entries(profile.regions))if((profile.regionAliases[r]||[]).some(a=>has(text,a)))region=Math.max(region,w*5);
 for(const [e,w] of Object.entries(profile.trackedEntities))if(has(text,e))entity=Math.max(entity,w*10);
 if(profile.preferredSources.some(s=>norm(story.source).includes(norm(s))))quality=9;else quality=6;
 for(const t of profile.noiseTerms)if(has(text,t))noise-=10;
 const ageH=Math.max(0,(Date.now()-new Date(story.publishedAt||Date.now()).getTime())/36e5);const freshness=Math.max(0,15-Math.min(15,ageH/8));
 const academic=story.type==='paper'?8:0;const importance=Math.min(20,5+entity*.8+topic*.25+academic);const novelty=Math.min(15,freshness);const future=Math.min(5,(topic/30)*3+(entity/10)*2);
 const score=Math.max(0,Math.min(100,topic+importance+novelty+quality+entity+region+future+noise));
 return {...story,topics,score,why:why(story,topics,entity,academic)};
}
function why(s,topics,entity,academic){if(academic)return 'Primary research aligned with your science and technical interests.';if(entity>=8)return 'A tracked lab, company or technology is directly involved.';if(topics.includes('neuroscience')&&topics.includes('artificial_intelligence'))return 'It sits at the AI × neuroscience intersection you care about.';if(topics.length)return `Strong overlap with ${topics.slice(0,3).map(x=>x.replaceAll('_',' ')).join(', ')}.`;return 'Included for broader significance and discovery.'}
export function dedupe(stories){const seen=new Set();return stories.filter(x=>{const k=norm(x.title).replace(/[^a-z0-9 ]/g,'').split(/\s+/).slice(0,10).join(' ');if(seen.has(k))return false;seen.add(k);return true;});}
export function digest(stories,profile,newsmarks=[]){const marked=new Set(newsmarks.map(x=>x.id));const ranked=dedupe(stories.map(s=>rankStory(s,profile))).sort((a,b)=>b.score-a.score).map(x=>({...x,newsmarked:marked.has(x.id)}));const top=ranked.slice(0,10);const cats=[['AI & Compute',['artificial_intelligence','computer_science']],['Startups & VC',['startups','venture_capital']],['Science & Neuroscience',['science','neuroscience']],['Hardware & Robotics',['hardware','semiconductors','robotics']],['Math & Ideas',['mathematics']]];
 const sections=cats.map(([name,ts])=>({name,items:ranked.filter(x=>x.topics.some(t=>ts.includes(t))).slice(0,8)})).filter(x=>x.items.length);
 const deep=ranked.filter(x=>x.type==='paper'||/research|paper|study|report/i.test(x.title)).slice(0,5);
 const outside=ranked.filter(x=>x.score>=30&&!x.topics.length).slice(0,4);
 return {top,sections,deep,outside,ranked};}

File: rank.py
import json, math, re
from datetime import datetime, timezone

TOKEN=re.compile(r"[a-z0-9+.-]+")

def toks(s): return set(TOKEN.findall((s or "").lower()))

def phrase_hit(text, phrase):
    p=phrase.lower().strip()
    if len(p) <= 3: return p in toks(text)
    return p in text.lower()

def score_story(story, profile):
    text=f"{story.get('title','')} {story.get('summary','')}"
    topic_score=0.0; hit_topics=[]
    for topic, weight in profile.get("topics",{}).items():
        aliases=[topic.replace("_"," ")] + profile.get("topic_aliases",{}).get(topic,[])
        if any(phrase_hit(text,a) for a in aliases):
            topic_score += 30*float(weight)/max(1, len([x for x in profile.get('topics',{}) if profile['topics'][x] >= .7]))
            hit_topics.append(topic)
    topic_score=min(30, topic_score*3.2)

    entity_score=0; entities=[]
    for ent, weight in profile.get("tracked_entities",{}).items():
        if phrase_hit(text, ent):
            entity_score += 10*float(weight)
            entities.append(ent)
    entity_score=min(10, entity_score)

    src=story.get("source","").lower()
    source_quality=5
    for preferred in profile.get("preferred_sources",[]):
        if preferred.lower() in src: source_quality=10; break
    if "arxiv" in src or "nature" in src: source_quality=max(source_quality,9)

    title=(story.get("title") or "").lower(); summary=(story.get("summary") or "").lower()
    importance=7
    importance_terms={"launch":2,"release":2,"acquire":3,"acquisition":3,"breakthrough":5,"regulation":3,"funding":1,"raises":1,"chip":2,"model":2,"research":2,"trial":3,"discovery":3,"open source":2}
    for k,v in importance_terms.items():
        if k in title: importance+=v
    importance=min(20,importance)

    novelty=7
    if any(k in title for k in ["new ","first ","unveils","launches","discovers","demonstrates","introduces"]): novelty+=4
    if any(k in summary for k in ["first time","novel","new method","new architecture"]): novelty+=2
    novelty=min(15,novelty)

    depth=1
    if "arxiv" in src or any(k in text.lower() for k in ["paper","researchers","study","dataset","benchmark","theorem"]): depth=5

    geography=3
    low=text.lower()
    for region,w in profile.get("regions",{}).items():
        aliases=profile.get("region_aliases",{}).get(region,[region])
        if any(a.lower() in low for a in aliases): geography=max(geography,5*float(w))
    geography=min(5,geography)

    future=2
    if any(k in low for k in ["infrastructure","semiconductor","compute","neuroscience","foundation model","robot","quantum","biotech","energy"]): future=5

    penalty=0
    noise_terms=profile.get("noise_terms",["celebrity","sports betting","daily horoscope"])
    for term in noise_terms:
        if term.lower() in low: penalty -= 12
    if any(k in title for k in ["price target","stock rises","stock falls","shares rise","shares fall"]): penalty-=10
    if "raises" in title and topic_score < 12 and entity_score == 0: penalty-=5

    total=topic_score+importance+novelty+source_quality+entity_score+depth+geography+future+penalty
    # freshness bonus, capped and forgiving of missing dates
    try:
        dt=datetime.fromisoformat((story.get("published_at") or "").replace("Z","+00:00"))
        age=max(0,(datetime.now(timezone.utc)-dt.astimezone(timezone.utc)).total_seconds()/3600)
        total += max(0, 5-age/12)
    except Exception: pass
    total=max(0,min(100,total))
    detail={"topic":round(topic_score,1),"importance":importance,"novelty":novelty,"source_quality":source_quality,"entity":round(entity_score,1),"depth":depth,"geography":round(geography,1),"future_significance":future,"noise_penalty":penalty}
    why = _why(hit_topics, entities, story)
    return round(total,1), detail, hit_topics, entities, why

def _why(topics, entities, story):
    if entities and topics: return f"Tracks {entities[0]} and intersects with your interest in {topics[0].replace('_',' ')}."
    if entities: return f"You asked Twin to track {entities[0]} closely."
    if topics: return f"Strong match for your {topics[0].replace('_',' ')} interests."
    if "arxiv" in story.get("source","").lower(): return "Primary research that may be worth scanning before it reaches mainstream coverage."
    return "Selected for significance and potential relevance outside your usual feed."
